scipy/sparse/linalg/propack.py

```diff
@@ -0,0 +1,328 @@
+"""
+Python wrapper for PROPACK
+--------------------------
+
+PROPACK is a collection of Fortran routines for iterative computation
+of partial SVDs of large matrices or linear operators.
+
+Based on BSD licensed pypropack project:
+  http://github.com/jakevdp/pypropack
+  Author: Jake Vanderplas <vanderplas@astro.washington.edu>
+
+PROPACK source is BSD licensed, and available at
+  http://soi.stanford.edu/~rmunk/PROPACK/
+"""
+
+__all__ = ['svdp']
+
+import numpy as np
+
+from scipy._lib._util import check_random_state
+from scipy.sparse.linalg import aslinearoperator
+from scipy.linalg import LinAlgError
+
+from ._propack import _spropack
+from ._propack import _dpropack
+from ._propack import _cpropack
+from ._propack import _zpropack
+
+
+_lansvd_dict = {
+    'f': _spropack.slansvd,
+    'd': _dpropack.dlansvd,
+    'F': _cpropack.clansvd,
+    'D': _zpropack.zlansvd,
+}
+
+
+_lansvd_irl_dict = {
+    'f': _spropack.slansvd_irl,
+    'd': _dpropack.dlansvd_irl,
+    'F': _cpropack.clansvd_irl,
+    'D': _zpropack.zlansvd_irl,
+}
+
+_which_converter = {
+    'LM': 'L',
+    'SM': 'S',
+}
+
+class _AProd:
+    """
+    Wrapper class for linear operator
+
+    The call signature of the __call__ method matches the callback of
+    the PROPACK routines.
+    """
+    def __init__(self, A):
+        try:
+            self.A = aslinearoperator(A)
+        except TypeError:
+            self.A = aslinearoperator(np.asarray(A))
+
+    def __call__(self, transa, m, n, x, y, sparm, iparm):
+        if transa == 'n':
+            y[:] = self.A.matvec(x)
+        else:
+            y[:] = self.A.rmatvec(x)
+
+    @property
+    def shape(self):
+        return self.A.shape
+
+    @property
+    def dtype(self):
+        try:
+            return self.A.dtype
+        except AttributeError:
+            return self.A.matvec(np.zeros(self.A.shape[1])).dtype
+
+
+def svdp(A, k, which='LM', irl_mode=True, kmax=None,
```

Comment:
Oops, weren't we going to make `_svdp` private for now?

Reply:
Yeah, we should.  `svdp` should be renamed to `_svdp` and `propack.py` should be renamed to `_propack.py`

Reply:
Ok, are you doing that or you want me to?

Reply:
I can do it -- just rebuilding now to make sure nothing broke from the renaming

Reply:
Tell you what - if you can do that, I can try rewriting those Fortran file loaders in Python now that I know that it's just csc format.

Reply:
Okay, I think this is done.

Another question is where the submodule should live.  Boost lives in `scipy._lib` because it's a general library, but the UNU.RAN wrapper (when it comes) is also currently targeting that directory such that submodules are grouped together.  The way we have it now, the PROPACK submodule lives next to its wrapper (which is the way we've historically vendored code).  I don't really have an opinion either way -- thoughts @mdhaber @rgommers?

Reply:
I agree, if it's a submodule that will only be used from one place, that's where it should live (just like if we would vendor the code). Boost is probably a bit of an exception; it's somewhat likely we'll use it directly in multiple scipy submodules.